src/cpufreq.c

```diff
@@ -27,39 +45,212 @@
+	// frequencies.
+	//
+	// Limitation of this approach: if at the time of scan a CPU
+	// that has some frequencies not covered by other CPUs is
+	// switched off and the cpufreq stat for that CPU is not
+	// available, the corresponding frequency would be missing
+	// from the list.
+	if (num_cpu > 0 && report_distribution && !report_by_cpu)
+	{
+		FILE *fp;
+		char *fields[2];
+		char buffer[256];
+		int numfields;
+		long int hertz;
+		int found;
+
+		// initialize list of frequencies
+		hertz_all_cpus = NULL;
+		hertz_size = 0;
+
+		// scan all CPUs for available frequencies
+		for (size_t i=0; i < num_cpu; ++i)
+		{
+			// To get frequencies, we scan the stats file
+			// that is used later as well. Format of the
+			// file is:
+			//
+			// hertz1 time_in_10ms
+			// hertz2 time_in_10ms
+			// ...
+			//
+			// At this stage, only hertz is used.
+			status = ssnprintf (filename, sizeof (filename),
+					    "/sys/devices/system/cpu/cpu%zu/cpufreq/stats/time_in_state",
+					    i);
+
+			if ((status < 1) || ((size_t)status >= sizeof (filename)))
+			{
+				INFO ("cpufreq plugin: error in ssnprintf");
+				break;
+			}
+
+			if ((fp = fopen (filename, "r")) == NULL)
+				// cpu could be switched off, just take the next one
+				continue;
+
+			while (fgets(buffer, sizeof(buffer), fp) != NULL)
+			{
+				numfields = strsplit (buffer, fields, STATIC_ARRAY_SIZE(fields));
+
+				if (numfields != 2) // supported format contains 2 fields
+				{
+					// if its an empty line, just ignore this line
+					if (numfields == 0) continue;
+
+					// unsupported format, report error and unregister
+					WARNING ("cpufreq plugin: wrong number of items on a line in %s", filename);
+					fclose (fp);
+					plugin_unregister_read ("cpufreq");
+					return (0);
+				}
+
+				hertz = atol (fields[0]);
+
+				// do we have this value already?
+				found = 0;
+				for (size_t j=0; j < hertz_size && !found; ++j)
+					if (hertz == hertz_all_cpus[j])
+						found = 1;
+
+				// this is a new frequency. add this
+				// to the list at the end. when
+				// scanning for stats later,
+				// frequencies should come in the same
+				// order, so the sorting is not needed
+				// as such (see read function below).
+				if (!found)
+				{
+					hertz_size += 1;
+					hertz_all_cpus = (long int*)realloc (hertz_all_cpus, hertz_size * sizeof(long int));
+					if (hertz_all_cpus == NULL)
+					{
+						INFO ("cpufreq plugin: realloc failed for hertz_all_cpus size %zu", hertz_size);
+						fclose (fp);
+						plugin_unregister_read ("cpufreq");
+						return (0);
+					}
+					hertz_all_cpus[ hertz_size-1 ] = hertz;
+				}
+			}
+
+			fclose (fp);
+		}
+
+		if (hertz_size == 0)
+		{
+			INFO ("cpufreq plugin: cannot find any frequencies in the stats");
+			plugin_unregister_read ("cpufreq");
+			return (0);
+		}
+
+		// Allocate memory for time array that will be used in
+		// read function.
+		time_all_cpus = malloc (hertz_size * sizeof(*time_all_cpus));
```

Comment:
time_all_cpus is a pointer (see line 66). initially its value is null, but since sizeof is evaluated by compiler it returns the size of the var type its pointing to.